### PR TITLE
Make test method's request

### PR DESCRIPTION
**Pull request by mehdi-fathi**
Make test method's request and fix some of problems in Request.php

### CODE REVIEW

File: src/Http/Client/Request.php [hunk 1]
```diff
@@ -253,5 +253,5 @@ public function body($body = null)
     }
 }
 
-// @deprecated Add backwards compat alias.
+// @deprecated Add backwards compact alias.
```
Comment (robertpustulka) [inline src/Http/Client/Request.php]: I think it's short for backwards compat*ibility*

File: tests/TestCase/Http/Client/RequestTest.php [hunk 2]
```diff
@@ -292,14 +341,14 @@ public function testCookie()
     public function testVersion()
     {
         $request = new Request();
-        $result = $request->version('1.0');
+        $request->version('1.0');
         $this->assertSame($request, $request, 'Should return self');
```
Comment (robertpustulka) [inline tests/TestCase/Http/Client/RequestTest.php]: This should be fixed to assert against `$result`

File: src/Http/Client/Request.php (hunk 1 above)
Comment (markstory) [inline src/Http/Client/Request.php]: It is.

File: src/Http/Client/Request.php [hunk 3]
```diff
@@ -9,7 +9,7 @@
  * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
  * @link          https://cakephp.org CakePHP(tm) Project
  * @since         3.0.0
- * @license       https://opensource.org/licenses/mit-license.php MIT License
+ * @license       https://www.opensource.org/licenses/mit-license.php MIT License
```
Comment (markstory) [inline src/Http/Client/Request.php]: You don't need the www on the https URLs.

File: tests/TestCase/Http/Client/RequestTest.php [hunk 4]
```diff
@@ -37,7 +48,44 @@ public function testConstructorStringData()
...
+        $headers = [
+            'Content-Type' => 'application/json',
+            'Authorization' => 'Bearer valid-token',
+        ];
+        $data = ['a' => 'b', 'c' => 'd'];
+        $request = new Request('http://example.com', 'POST', $headers, json_encode($data));
+        $this->assertContains($request->getMethod(), $this->methods);
+        $this->assertEquals('http://example.com', $request->url());
+        $this->assertEquals('application/json', $request->getHeaderLine('Content-Type'));
+        $this->assertEquals(json_encode($data), $request->body());
```
Comment (markstory) [inline tests/TestCase/Http/Client/RequestTest.php]: This looks like a test that could benefit from a [dataprovider](https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers).

**Review by markstory — CHANGES_REQUESTED** [review]

File: tests/TestCase/Http/Client/RequestTest.php (hunk 2 above)
Comment (mehdi-fathi) [inline tests/TestCase/Http/Client/RequestTest.php]: Yes exacly

File: tests/TestCase/Http/Client/RequestTest.php [hunk 5]
```diff
@@ -37,10 +48,44 @@ public function testConstructorStringData()
...
+     * @param array|string|null $data The request body to use.
+     * @param string $method The HTTP method to use.
+     *
+     * @dataProvider additionProvider
+     */
+    public function testAdd(array $headers, $data, $method)
+    {
+        $request = new Request('http://example.com', $method, $headers, json_encode($data));
+
+        $this->assertContains($request->getMethod(), $this->methods);
```
Comment (markstory) [inline tests/TestCase/Http/Client/RequestTest.php]: Wouldn't it be more accurate to check that the method matches exactly? Right now the test would continue to pass even if the request method was transposed to another valid method that wasn't requested.

Comment (mehdi-fathi) [inline tests/TestCase/Http/Client/RequestTest.php]: It make no difference.but you are right assertcontains isn't suitable here.

File: src/Http/Client/Request.php (hunk 1 above)
Comment (ravage84) [inline src/Http/Client/Request.php]: I would recommend to write out the word compatibility in the future.

**Pull request merged**